### PR TITLE
Document labeled metrics types as separate individual types (bug 1553445) [ci skip]

### PR DESCRIPTION
**Pull request by georgf**
This clearly calls out the labeled metric types as separate, individual types in the documentation.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by georgf** [comment]
I'd like to get a first set of eyes on it, there are some clearly marked TODOs in it still.
- If you can think of better examples, let me know.
- For `labeled_string` & `labeled_boolean` i wonder if we have clear & good use-cases. If not, maybe we should disable this types for now and wait for use-cases to come up?

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 1]
```diff
@@ -80,15 +80,28 @@ Views.loginOpened...
 There are different metrics to choose from, depending on what you want to achieve:
 
 * [Events](#Events): Records events e.g. individual occurences of user actions, say every time a view was open and from where.
-* [Booleans](#Booleans): Records a single truth value, for example "is a11y enabled?"
+* [Booleans](#Booleans): Records a single truth value, for example "is a11y enabled?".
+* [Labeled booleans](#Labeled-booleans): Records multiple truth values, for example which URI schemes were used or not - `"http"`, `"https"`, `"about"`.
 * [Strings](#Strings): Records a single Unicode string value, for example the name of the OS.
-* [String Lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines.
+* [Labeled strings](#Labeled-strings): Records multiple Unicode string values. **TODO: what's the use-case here? Should we block these for now?**
```
Comment (travis79) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: Labeled strings seem to overlap quite a bit with string lists.  I'd be okay with keeping them out until a strong use-case comes up.

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 2]
```diff
@@ -80,15 +80,28 @@ Views.loginOpened...
...
+* [String Lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines **TODO - i can't think of a good example here**.
 * [Counters](#Counters): Used to count how often something happens, for example, how often a certain button was pressed.
+* [Labeled counters](#Labeled-counters): Used to count how often something happens, for example which kind of crashes occured - `"uncaught_exception"` or `"native_code_crash"`.
```
Comment (travis79) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: 😄 Good example!

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 3]
```diff
@@ -293,6 +331,25 @@ assertTrue(Controls.refreshPressed.testHasValue())
 assertEquals(6, Controls.refreshPressed.testGetValue())
 ```
 
+## Labeled Counters
+
+Labeled counters are used to record different related counts that should sum up to a total.
+
+For example, you may want to record a count of different types of crashes for your Android application, native code crashes and uncaught exceptions:
```
Comment (travis79) [inline components/service/glean/docs/metrics/adding-new-metrics.md]:
```suggestion
For example, you may want to record a count of different types of crashes for your Android application, such as native code crashes and uncaught exceptions:
```

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 4]
```diff
@@ -368,6 +425,25 @@ assertTrue(Auth.loginTime.testHasValue())
 assertTrue(Auth.loginTime.testGetValue() > 0)
 ```
 
+## Labeled Timespans
+
+Used to measure how much time is spent in a set of related tasks.
+
+For example to record the time spent in different stages in a login process:
```
Comment (travis79) [inline components/service/glean/docs/metrics/adding-new-metrics.md]:
```suggestion
For example, to record the time spent in different stages in a login process:
```

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 5]
```diff
@@ -175,6 +188,27 @@ assertTrue(Flags.a11yEnabled.testHasValue())
 assertTrue(Flags.a11yEnabled.testGetValue())
 ```
 
+## Labeled Booleans
+
+Used to record multiple related truth values
```
Comment (travis79) [inline components/service/glean/docs/metrics/adding-new-metrics.md]:
```suggestion
Used to record multiple related truth values.
```

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 6]
```diff
@@ -80,15 +80,28 @@ Views.loginOpened...
...
 * [Counters](#Counters): Used to count how often something happens, for example, how often a certain button was pressed.
+* [Labeled counters](#Labeled-counters): Used to count how often something happens, for example which kind of crashes occured - `"uncaught_exception"` or `"native_code_crash"`.
 * [Timespans](#Timespans): Used to measure how much time is spent in a single task.
+* [Labeled timespans](#Labeled-timespans): Used to measure how much time is spent in a set of related tasks, for example different stages in a login process - the time taken for `"fill_form"`, `"auth_with_server"`, `"load_next_view"`.
 * [Timing Distributions](#Timing-Distributions): Used to record the distribution of multiple time measurements.
 * [Datetimes](#Datetimes): Used to record an absolute date and time, such as the time the user first ran the application.
 * [UUIDs](#UUIDs): Used to record universally unique identifiers (UUIDs), such as a client ID.
-* [Labeled Metrics](#Labeled-metrics): Used to record multiple metrics of the same type under different string labels, say every time you want to get a count of different error types in one metric.
+
+## Labeled metrics
```
Comment (travis79) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: Would it be useful to call out the restrictions on labels here, such as they can only be dotted.snake_case, or the max length?

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 7]
```diff
@@ -80,15 +80,28 @@ Views.loginOpened...
 There are different metrics to choose from, depending on what you want to achieve:
 
 * [Events](#Events): Records events e.g. individual occurences of user actions, say every time a view was open and from where.
-* [Booleans](#Booleans): Records a single truth value, for example "is a11y enabled?"
+* [Booleans](#Booleans): Records a single truth value, for example "is a11y enabled?".
+* [Labeled booleans](#Labeled-booleans): Records multiple truth values, for example which URI schemes were used or not - `"http"`, `"https"`, `"about"`.
 * [Strings](#Strings): Records a single Unicode string value, for example the name of the OS.
-* [String Lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines.
+* [Labeled strings](#Labeled-strings): Records multiple Unicode string values. **TODO: what's the use-case here? Should we block these for now?**
+* [String Lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines **TODO - i can't think of a good example here**.
```
Comment (Dexterp37) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: "What is the list of names for the accessibility services available on the device?"

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 1 above)
Comment (Dexterp37) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: Let's file a bug about this, please.

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 8]
```diff
@@ -80,15 +80,28 @@ Views.loginOpened...
...
 * [Timespans](#Timespans): Used to measure how much time is spent in a single task.
+* [Labeled timespans](#Labeled-timespans): Used to measure how much time is spent in a set of related tasks, for example different stages in a login process - the time taken for `"fill_form"`, `"auth_with_server"`, `"load_next_view"`.
 * [Timing Distributions](#Timing-Distributions): Used to record the distribution of multiple time measurements.
 * [Datetimes](#Datetimes): Used to record an absolute date and time, such as the time the user first ran the application.
 * [UUIDs](#UUIDs): Used to record universally unique identifiers (UUIDs), such as a client ID.
-* [Labeled Metrics](#Labeled-metrics): Used to record multiple metrics of the same type under different string labels, say every time you want to get a count of different error types in one metric.
+
+## Labeled metrics
+
+There are two types of metrics listed above - *labeled* and *unlabeled* metrics. If a metric is *labeled*, it means that for a single metric entry you define in `metrics.yaml`, you can record into multiple metrics under the same name, each identified by a different string label.
```
Comment (Dexterp37) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: ", each of the same type and identified by..."

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 6 above)
Comment (Dexterp37) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: I'm not sure we should get into this level of detail: builds will fail anyway if we do something wrong.

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 8 above)
Comment (Dexterp37) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: I'd dig into the difference of the dynamic/static labels too, as shown in the [design doc](https://docs.google.com/document/d/186GI_2ljdqWqWTBk5P3QjHD3BwpMyd8kRayRAuz1-dk/edit#heading=h.z4j96kxogool)

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 9]
```diff
@@ -293,6 +331,25 @@ assertTrue(Controls.refreshPressed.testHasValue())
...
+  type: labeled_counter
+  description: >
+    Counts the number of crashes that occur in the application. ...
+  labels:
+    - uncaught_exception
+    - native_code_crash
+  ...
+```
+
+**TODO: test example here**
```
Comment (Dexterp37) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: Would it be redundant with [the other docs](https://github.com/mozilla-mobile/android-components/blob/master/components/service/glean/docs/metrics/testing-metrics.md) ?

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 10]
```diff
@@ -175,6 +188,27 @@ assertTrue(Flags.a11yEnabled.testHasValue())
 assertTrue(Flags.a11yEnabled.testGetValue())
 ```
 
+## Labeled Booleans
+
+Used to record multiple related truth values
+
+For example you may want to find which URI schemes are used in your application, like `"http"`, `"https"`, `"about"`:
+
+**TODO: What's the typical use-cases here? Should we instead really do a kind of "label set" metric?**
```
Comment (Dexterp37) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: What do you mean by "label set"?

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 8 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: This section is just moved around in the doc.
I'll file a follow-up on that.

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 10 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: A set of unique string values (vs. a string list with non-uniques). E.g.:
```
val x: StringSetMetric
x.add("http")
x.add("http")
x.add("about")
// x now contains: {"http", "about"}
```

Comment (Dexterp37) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: Ah, good point. Can you please file a bug?

**Comment by Dexterp37** [comment]
@georgf can you kindly file a bug for backporting these changes to glean-core?

**Comment by mdboom** [comment]
> @georgf can you kindly file a bug for backporting these changes to glean-core?

As I've brought docs over to glean-core, I've been very careful in the user docs to make sure it applies equally to glean-ac and glean-core.  So at some point, maybe even before the code migration, we can probably replace all the docs in glean-ac with a link to the glean-core docs.

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 1 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: https://bugzilla.mozilla.org/show_bug.cgi?id=1557349

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 7 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: I wonder if that should really be a string set then.
I'll file a bug on that.

Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: https://bugzilla.mozilla.org/show_bug.cgi?id=1557353

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 1 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: Per the latest comments, this is in use by sync.
We should keep it for now and defer the discussion.
https://bugzilla.mozilla.org/show_bug.cgi?id=1557349#c2

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 10 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: https://bugzilla.mozilla.org/show_bug.cgi?id=1557353

File: components/service/glean/docs/metrics/adding-new-metrics.md (hunk 9 above)
Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: I think each metric type section should stand on it's own.
We do have a "what is it for?", "what do i put in metrics.yaml?", "how do i use it from my code?", and finally this section, "how do i use it in tests?".

**Comment by georgf** [comment]
This is up for review now.

File: components/service/glean/docs/metrics/adding-new-metrics.md [hunk 11]
```diff
@@ -80,15 +80,27 @@ Views.loginOpened...
 There are different metrics to choose from, depending on what you want to achieve:
 
 * [Events](#Events): Records events e.g. individual occurences of user actions, say every time a view was open and from where.
```
Comment (Dexterp37) [inline components/service/glean/docs/metrics/adding-new-metrics.md]: I'm not sure where it should fit, but I now realize that we really want to express our opinion with respect to metric names: maybe we should explain why `events.button_event` isn't a good naming scheme (e.g. event is redundant as a category name and button_event doesn't really explain what the event is about). Any chance you can tackle this here on file a follow-up for it?

Comment (georgf) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: That sounds orthogonal to the PR here.
Can you file a follow-up on what your intent there is @Dexterp37  ?

Comment (Dexterp37) [reply components/service/glean/docs/metrics/adding-new-metrics.md]: Filed this. https://bugzilla.mozilla.org/show_bug.cgi?id=1557700

**Pull request merged**